nodl_python/nodl/_parsing/_v1/_parsing.py

```diff
@@ -0,0 +1,119 @@
+    action_type = element.get('type')
+
+    server = get_bool_attribute(element, 'server')
+    client = get_bool_attribute(element, 'client')
+    if not (server or client):
+        raise NoNodeInterfaceError(f'Action <{name}> is neither server nor client.', element)
+
+    policy = _parse_qos(element.find('qos'))
+
+    return Action(name=name, action_type=action_type, server=server, client=client, qos=policy)
+
+
+def _parse_parameter(element: etree._Element) -> Parameter:
+    """Parse a NoDL parameter from an xml element."""
+    return Parameter(name=element.get('name'), parameter_type=element.get('type'))
+
+
+def _parse_service(element: etree._Element) -> Service:
+    """Parse a NoDL service from an xml element."""
+    name = element.get('name')
+    service_type = element.get('type')
+
+    server = get_bool_attribute(element, 'server')
+    client = get_bool_attribute(element, 'client')
+    if not (server or client):
+        raise NoNodeInterfaceError(f'Service <{name}> is neither server nor client.', element)
+
+    policy = _parse_qos(element.find('qos'))
+
+    return Service(name=name, service_type=service_type, server=server, client=client, qos=policy,)
+
+
+def _parse_topic(element: etree._Element) -> Topic:
+    """Parse a NoDL topic from an xml element."""
+    name = element.get('name')
+    message_type = element.get('type')
+
+    publisher = get_bool_attribute(element, 'publisher')
+    subscription = get_bool_attribute(element, 'subscription')
+    if not (publisher or subscription):
+        raise NoNodeInterfaceError(
+            f'Topic <{name}> is neither publisher nor subscription.', element
+        )
+
+    policy = _parse_qos(element.find('qos'))
+
+    return Topic(
+        name=name,
+        message_type=message_type,
+        publisher=publisher,
+        subscription=subscription,
+        qos=policy,
+    )
+
+
+def _parse_nodes(interface: etree._Element) -> List[Node]:
+    """Parse the nodes contained in an interface element and return a list."""
+    node_elements = [child for child in interface if child.tag == 'node']
+    return [_parse_node(node) for node in node_elements]
+
+
+def _parse_node(node: etree._Element) -> Node:
+    """Parse a NoDL node and all the elements it contains from an xml element."""
+    name = node.attrib['name']
+
+    actions = []
+    parameters = []
+    services = []
+    topics = []
+
+    for child in node:
+        if child.tag == 'action':
+            actions.append(_parse_action(child))
+        if child.tag == 'parameter':
+            parameters.append(_parse_parameter(child))
+        if child.tag == 'service':
+            services.append(_parse_service(child))
+        if child.tag == 'topic':
+            topics.append(_parse_topic(child))
+    return Node(
+        name=name, actions=actions, parameters=parameters, services=services, topics=topics
+    )
+
+
+def _validate_and_parse(interface: etree._Element) -> List[Node]:
```

Comment:
This should probably be the entry point into this package. As a result it should probably be public. I'd also just call it `parse`.

Reply:
Fixed in db8e13c

Reply:
Not quite, still needs to be public and called `parse` :wink: .

Reply:
36f9562

Reply:
There we go!